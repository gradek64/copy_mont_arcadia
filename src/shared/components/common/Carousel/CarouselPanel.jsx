import React, { Component } from 'react'
import classNames from 'classnames'
import Hammer from 'react-hammerjs'
import PropTypes from 'prop-types'
import { path } from 'ramda'

export default class CarouselPanel extends Component {
  static propTypes = {
    name: PropTypes.string.isRequired,
    carousel: PropTypes.object.isRequired,
    assets: PropTypes.arrayOf(PropTypes.object).isRequired,
    backCarousel: PropTypes.func.isRequired,
    forwardCarousel: PropTypes.func.isRequired,
    mode: PropTypes.string,
    setCurrentItemReference: PropTypes.func,
    currentItemReferencePath: PropTypes.func,
    handleSwipe: PropTypes.func,
    arrowColor: PropTypes.string,
    className: PropTypes.string,
  }

  static defaultProps = {
    className: '',
  }

  constructor(props) {
    super(props)
    this.state = { shouldAnimate: false }
  }

  componentDidMount() {
    const {
      name,
      setCurrentItemReference,
      currentItemReferencePath,
    } = this.props
    if (currentItemReferencePath) {
      const currentIndex = path(['carousel', name, 'current'], this.props)
      setCurrentItemReference(
        name,
        currentItemReferencePath(this.props.assets[currentIndex])
      )
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const {
      name,
      setCurrentItemReference,
      currentItemReferencePath,
    } = this.props

    const nextIndex = path(['carousel', name, 'current'], nextProps)
    const currentIndex = path(['carousel', name, 'current'], this.props)

    if (nextIndex !== currentIndex) {
      this.setState({ shouldAnimate: true })
      if (currentItemReferencePath) {
        setCurrentItemReference(
          name,
          currentItemReferencePath(nextProps.assets[nextIndex])
        )
      }
    }
  }

  renderControls() {
    const { name, arrowColor, backCarousel, forwardCarousel } = this.props

    return [
      <button
        key="left"
        className={classNames('Carousel-arrow', 'Carousel-arrow--left')}
        onClick={() => backCarousel(name)}
        aria-hidden="true"
        tabIndex="-1"
        style={{ backgroundColor: arrowColor }}
      />,
      <button
        key="right"
        className={classNames('Carousel-arrow', 'Carousel-arrow--right')}
        onClick={() => forwardCarousel(name)}
        aria-hidden="true"
        tabIndex="-1"
        style={{ backgroundColor: arrowColor }}
      />,
    ]
  }

  render() {
    const { assets, carousel, mode, name, handleSwipe, className } = this.props

    if (!carousel[name]) return null

    const { current, previous, direction, initialIndex } = carousel[name]

    const content = (
      <div className="Carousel-items" aria-hidden="true">
        <ul className="Carousel-list" tabIndex="-1">
          {assets.map((asset, i) => {
            const selected = current === i
            const isFirst = i === initialIndex
            const selectedClass = selected
              ? 'selected'
              : previous === i
                ? 'previous'
                : 'inactive'

            const itemClass = () => {
              if (!this.state.shouldAnimate && isFirst) {
                return 'Carousel-initialItem'
              }
              return `Carousel-${direction}Item`
            }

            return (
              <li // eslint-disable-line jsx-a11y/no-static-element-interactions
                key={`i${name}${i}`} // eslint-disable-line react/no-array-index-key
                ref={`CarouselItem${i}`}
                className={`Carousel-item foo-${i} is-${selectedClass} ${itemClass()}`}
              >
                <div>{asset}</div>
              </li>
            )
          })}
        </ul>
      </div>
    )

    return (
      <div // eslint-disable-line jsx-a11y/no-static-element-interactions
        className={`Carousel${mode ? ` Carousel--${mode}` : ''} ${className}`}
        ref={(Carousel) => {
          this.CarouselRef = Carousel
        }}
      >
        <Hammer onSwipe={handleSwipe} direction="DIRECTION_HORIZONTAL">
          {content}
        </Hammer>
        {assets.length > 1 && this.renderControls()}
      </div>
    )
  }
}
