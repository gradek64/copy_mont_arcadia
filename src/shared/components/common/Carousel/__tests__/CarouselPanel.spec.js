import testComponentHelper from 'test/unit/helpers/test-component'
import CarouselPanel from '../CarouselPanel'
import React from 'react'
import { path } from 'ramda'

describe('@<CarouselPanel />', () => {
  const initialProps = {
    name: 'miniProductCarousel0',
    carousel: {
      miniProductCarousel0: {
        direction: 'right',
        previous: -1,
        current: 0,
        max: 6,
        zoom: 1,
        panX: 0,
        panY: 0,
        currentItemReference: undefined,
        initialIndex: 0,
      },
    },
    assets: [<div />, <div />, <div />],
    touchEnabled: false,
    backCarousel: jest.fn(),
    forwardCarousel: jest.fn(),
    className: 'Carousel--panel',
  }

  const renderComponent = testComponentHelper(CarouselPanel)

  describe('@renders', () => {
    it('in default state', () => {
      expect(renderComponent(initialProps).getTree()).toMatchSnapshot()
    })
    it('`null` if `initialProps.name` is not found in `initialProps.carousel`', () => {
      expect(
        renderComponent({
          ...initialProps,
          carousel: {},
        }).getTree()
      ).toBe('')
    })
    it('renders with correct `Carousel--panel` class', () => {
      const { wrapper } = renderComponent(initialProps)
      expect(wrapper.hasClass('Carousel--panel')).toBe(true)
    })
    it('first Carousel-item should have is-selected class', () => {
      const { wrapper } = renderComponent(initialProps)
      expect(
        wrapper
          .find('.Carousel-item')
          .at(0)
          .hasClass('is-selected')
      ).toBe(true)
    })
    it('controls do not render when `assets` prop is empty array', () => {
      const { wrapper } = renderComponent({
        ...initialProps,
        assets: [],
      })
      expect(wrapper.find('.Carousel-arrow--left').length).toBe(0)
      expect(wrapper.find('.Carousel-arrow--right').length).toBe(0)
    })

    it('should adjust the initial carousel image according to the initial index', () => {
      expect(
        renderComponent({
          ...initialProps,
          carousel: {
            ...initialProps.carousel,
            miniProductCarousel0: {
              ...initialProps.carousel.miniProductCarousel0,
              initialIndex: 2,
            },
          },
        }).getTree()
      ).toMatchSnapshot()
    })
  })

  describe('@events', () => {
    describe('baseControls click', () => {
      let renderedComponent

      beforeEach(() => {
        renderedComponent = renderComponent(initialProps)
        jest.resetAllMocks()
      })
      it('should call `backCarousel` on left arrow click', () => {
        const { wrapper, instance } = renderedComponent
        expect(instance.props.backCarousel).not.toHaveBeenCalled()
        wrapper.find('.Carousel-arrow--left').simulate('click')
        expect(instance.props.backCarousel).toHaveBeenCalledTimes(1)
        expect(instance.props.backCarousel).toHaveBeenCalledWith(
          instance.props.name
        )
      })
      it('should call `forwardCarousel` on right arrow click', () => {
        const { wrapper, instance } = renderedComponent
        expect(instance.props.forwardCarousel).not.toHaveBeenCalled()
        wrapper.find('.Carousel-arrow--right').simulate('click')
        expect(instance.props.forwardCarousel).toHaveBeenCalledTimes(1)
        expect(instance.props.forwardCarousel).toHaveBeenCalledWith(
          instance.props.name
        )
      })
    })
    describe('Hammer swipe', () => {
      it('should call `handleSwipe`', () => {
        const { wrapper, instance } = renderComponent({
          ...initialProps,
          handleSwipe: jest.fn(),
        })
        expect(instance.props.handleSwipe).not.toHaveBeenCalled()
        wrapper.find('Hammer').simulate('swipe')
        expect(instance.props.handleSwipe).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('@lifecycle', () => {
    describe('@componentDidMount', () => {
      it('should call `setCurrentItemReference` should be called once with correct args', () => {
        const { instance } = renderComponent({
          ...initialProps,
          setCurrentItemReference: jest.fn(),
          currentItemReferencePath: path(['props', 'productId']),
          assets: [
            <div productId={123} />,
            <div productId={234} />,
            <div productId={345} />,
          ],
        })
        expect(instance.props.setCurrentItemReference).not.toHaveBeenCalled()
        instance.componentDidMount()
        expect(instance.props.setCurrentItemReference).toHaveBeenCalledTimes(1)
        expect(instance.props.setCurrentItemReference).toHaveBeenCalledWith(
          instance.props.name,
          instance.props.assets[0].props.productId
        )
      })

      it('setCurrentItemReference` should not be called if `currentItemReferencePath` is falsey', () => {
        const { instance } = renderComponent({
          ...initialProps,
          setCurrentItemReference: jest.fn(),
          currentItemReferencePath: null,
        })
        instance.componentDidMount()
        expect(instance.props.setCurrentItemReference).not.toHaveBeenCalled()
      })

      it('currentItemReferencePath` should called once with correct args', () => {
        const { instance } = renderComponent({
          ...initialProps,
          setCurrentItemReference: jest.fn(),
          currentItemReferencePath: jest.fn(),
          assets: [
            <div productId={123} />,
            <div productId={234} />,
            <div productId={345} />,
          ],
        })
        expect(instance.props.currentItemReferencePath).not.toHaveBeenCalled()
        instance.componentDidMount()
        expect(instance.props.currentItemReferencePath).toHaveBeenCalledTimes(1)
        expect(instance.props.currentItemReferencePath).toHaveBeenCalledWith(
          instance.props.assets[0]
        )
        expect(
          instance.props.currentItemReferencePath
        ).not.toHaveBeenCalledWith(instance.props.assets[1])
      })
    })

    describe('@UNSAFE_componentWillReceiveProps', () => {
      it('should call `setCurrentItemReference` should be called once with correct args', () => {
        const { instance } = renderComponent({
          ...initialProps,
          setCurrentItemReference: jest.fn(),
          currentItemReferencePath: path(['props', 'productId']),
          assets: [
            <div productId={123} />,
            <div productId={234} />,
            <div productId={345} />,
          ],
        })

        expect(instance.props.setCurrentItemReference).not.toHaveBeenCalled()

        const nextProps = {
          ...initialProps,
          carousel: {
            miniProductCarousel0: {
              direction: 'right',
              previous: 0,
              current: 1,
              max: 6,
              zoom: 1,
              panX: 0,
              panY: 0,
              currentItemReference: instance.props.assets[0].props.productId,
            },
          },
          assets: [
            <div productId={123} />,
            <div productId={234} />,
            <div productId={345} />,
          ],
        }

        instance.UNSAFE_componentWillReceiveProps(nextProps)

        expect(instance.props.setCurrentItemReference).toHaveBeenCalledTimes(1)
        expect(instance.props.setCurrentItemReference).toHaveBeenCalledWith(
          instance.props.name,
          instance.props.assets[1].props.productId
        )
      })
    })

    it('setCurrentItemReference` should not be called if `nextIndex !== currentIndex` is false', () => {
      const { instance } = renderComponent({
        ...initialProps,
        setCurrentItemReference: jest.fn(),
      })

      const nextProps = {
        ...initialProps,
      }

      instance.UNSAFE_componentWillReceiveProps(nextProps)

      expect(instance.props.setCurrentItemReference).not.toHaveBeenCalled()
    })
  })
})
