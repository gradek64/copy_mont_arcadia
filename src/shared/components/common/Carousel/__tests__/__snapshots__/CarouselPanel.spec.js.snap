// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@<CarouselPanel /> @renders in default state 1`] = `
<div
  className="Carousel Carousel--panel"
>
  <Hammer
    direction="DIRECTION_HORIZONTAL"
  >
    <div
      aria-hidden="true"
      className="Carousel-items"
    >
      <ul
        className="Carousel-list"
        tabIndex="-1"
      >
        <li
          className="Carousel-item foo-0 is-selected Carousel-initialItem"
          key="iminiProductCarousel00"
        >
          <div>
            <div />
          </div>
        </li>
        <li
          className="Carousel-item foo-1 is-inactive Carousel-rightItem"
          key="iminiProductCarousel01"
        >
          <div>
            <div />
          </div>
        </li>
        <li
          className="Carousel-item foo-2 is-inactive Carousel-rightItem"
          key="iminiProductCarousel02"
        >
          <div>
            <div />
          </div>
        </li>
      </ul>
    </div>
  </Hammer>
  <button
    aria-hidden="true"
    className="Carousel-arrow Carousel-arrow--left"
    key="left"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
    tabIndex="-1"
  />
  <button
    aria-hidden="true"
    className="Carousel-arrow Carousel-arrow--right"
    key="right"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
    tabIndex="-1"
  />
</div>
`;

exports[`@<CarouselPanel /> @renders should adjust the initial carousel image according to the initial index 1`] = `
<div
  className="Carousel Carousel--panel"
>
  <Hammer
    direction="DIRECTION_HORIZONTAL"
  >
    <div
      aria-hidden="true"
      className="Carousel-items"
    >
      <ul
        className="Carousel-list"
        tabIndex="-1"
      >
        <li
          className="Carousel-item foo-0 is-selected Carousel-rightItem"
          key="iminiProductCarousel00"
        >
          <div>
            <div />
          </div>
        </li>
        <li
          className="Carousel-item foo-1 is-inactive Carousel-rightItem"
          key="iminiProductCarousel01"
        >
          <div>
            <div />
          </div>
        </li>
        <li
          className="Carousel-item foo-2 is-inactive Carousel-initialItem"
          key="iminiProductCarousel02"
        >
          <div>
            <div />
          </div>
        </li>
      </ul>
    </div>
  </Hammer>
  <button
    aria-hidden="true"
    className="Carousel-arrow Carousel-arrow--left"
    key="left"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
    tabIndex="-1"
  />
  <button
    aria-hidden="true"
    className="Carousel-arrow Carousel-arrow--right"
    key="right"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
    tabIndex="-1"
  />
</div>
`;
